rest/api.go

```diff
@@ -64,19 +64,61 @@ func (h *handler) handleVacuum() error {
 }
 
 func (h *handler) handleFlush() error {
-	if bucket, ok := h.db.Bucket.(sgbucket.DeleteableBucket); ok {
+
+	if _, ok := h.db.Bucket.(sgbucket.DeleteableBucket); ok {
+
+		// If it's not a walrus bucket, don't allow flush unless the unsupported config is set
+		if !h.db.BucketSpec.IsWalrusBucket() {
+			if !h.db.DatabaseContext.AllowFlushNonCouchbaseBuckets() {
+				msg := "Flush not allowed on Couchbase buckets.  You must explicitly enable the ability to flush " +
+					"Couchbase buckets via the Unsupported option: api_endpoints/enable_couchbase_bucket_flush"
+				return fmt.Errorf(msg)
+			}
+		}
+
 		name := h.db.Name
 		config := h.server.GetDatabaseConfig(name)
+
+		// This needs to first call RemoveDatabase since flushing the bucket under Sync Gateway might cause issues.
 		h.server.RemoveDatabase(name)
-		err := bucket.CloseAndDelete()
+
+		// Create a bucket connection spec from the database config
+		spec, err := GetBucketSpec(config)
+		if err != nil {
+			return err
+		}
+
+		// Manually re-open a temporary bucket connection just for flushing purposes
+		tempBucketForFlush, err := db.ConnectToBucket(spec, nil)
+		if err != nil {
+			return err
+		}
+
+		// Flush the bucket (assuming it conforms to sgbucket.DeleteableBucket interface
+		if tempBucketForFlush, ok := tempBucketForFlush.(sgbucket.DeleteableBucket); ok {
+
+			// Flush
+			err := tempBucketForFlush.CloseAndDelete()
+			if err != nil {
+				return err
+			}
+
+		}
+
+		// Close the temporary connection to the bucket that was just for purposes of flushing it
+		tempBucketForFlush.Close()
```

Comment:
If we return an error from `tempBucketForFlush.CloseAndDelete()`, we'll never close the bucket. Could we defer the close after successfully connecting to the bucket instead?

Reply:
Fixed.  Perfect use of defer()!